Fix performance regression related to delayed events processing

anoadragon453 opened the pull request:
Hopefully fixes https://github.com/element-hq/synapse/issues/18925.

Carries out a number of optimisations to this bit of the codebase (note that on element.io we have about 7mil state deltas to process):

* Replace a `get_event` DB query per state delta with a single, smaller query for all state deltas.
* Bail out early if there are no delayed events in the DB to update.
* ~~Set the next "scheduled at" time to the minimum of all delayed events once, instead of doing so for every delta.~~ this ended up causing tests to fail, and has been reverted.

Highly recommended to review commit-by-commit.

MSC4140

### Pull Request Checklist

<!-- Please read https://element-hq.github.io/synapse/latest/development/contributing_guide.html before submitting your pull request -->

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://element-hq.github.io/synapse/latest/development/contributing_guide.html#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
  - Feel free to credit yourself, by adding a sentence "Contributed by @github_username." or "Contributed by [Your Name]." to the end of the entry.
* [x] [Code style](https://element-hq.github.io/synapse/latest/code_style.html) is correct (run the [linters](https://element-hq.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

anoadragon453 committed:
Add a storage function to pull out senders for event _ids

This is preferable to pulling out entire events and only fetching the `sender` field.

anoadragon453 committed:
Stop calling `get_event` and use new storage function

We also only call it once for all events in the batch, instead of per state delta.

anoadragon453 committed:
Only schedule next delayed event once

There's no need to do this for every event in the batch. Just find the min time for all delayed events.

anoadragon453 committed:
Bail out early if there are no delayed events to send

There are zero delayed events on matrix.org's database. We can save a
lot of time pulling out state deltas by first just checking if there are
any delayed events to actually deal with.

anoadragon453 committed:
newsfile

MadLittleMods commented on synapse/handlers/delayed_events.py:
```diff
@@ -215,10 +224,7 @@ async def _handle_state_deltas(self, deltas: List[StateDelta]) -> None:
-            if not event:
-                continue
-            sender = UserID.from_string(event.sender)
+            sender = UserID.from_string(event_id_and_sender_dict[delta.event_id])
```
Previously, we would skip over a delta if the event didn't exist. Now it looks like we will fail everything with a `KeyError`.

Which way is correct? We should also comment why it's okay to skip if that's the case.

MadLittleMods commented on synapse/handlers/delayed_events.py:
```diff
@@ -202,6 +224,15 @@ async def _handle_state_deltas(self, deltas: List[StateDelta]) -> None:
+            [delta.event_id for delta in deltas if delta.event_id is not None]
+        )
+
+        # Note: No need to batch as `get_current_state_deltas` will only ever
+        # return 100 rows at a time.
```
This is coming from https://github.com/element-hq/synapse/blob/6f9fab10890e9ad3afffb34fe40238b7e73155b0/synapse/storage/databases/main/state_deltas.py#L126

This sort of detail would be nice in the docstring of `get_current_state_deltas` (and the call chain there)

MadLittleMods commented on synapse/handlers/delayed_events.py:
```diff
@@ -146,10 +146,32 @@ async def process() -> None:
         )
 
     async def _unsafe_process_new_event(self) -> None:
+        room_max_stream_ordering = self._store.get_room_max_stream_ordering()
+
+        # Check that there are actually any delayed events to process. If not, bail early.
+        delayed_events_count = await self._store.get_count_of_delayed_events()
+        if delayed_events_count == 0:
```
This seems slightly sketchy since these two calls are happening in separate transactions.

But I guess we're assuming that if we get `room_max_stream_ordering` before, and then we check for any delayed events, we can assume there is none up to `room_max_stream_ordering`.

This wouldn't work if it was the other way around.

Trying to think if this also assumes that we're the only ones touching `delayed_events` :thinking: 

---

Overall, since we're adding a new `get_count_of_delayed_events()` check here anyway, we might as well craft it to return the `stream_ordering` as well so we don't even have to think about this.

anoadragon453 replied:
I'm not entirely sure how to get the latest events stream ordering in the same txn as `get_count_of_delayed_events()`, as this value appears to be [primarily kept in memory](https://github.com/element-hq/synapse/blob/40f077663057b5cc0888d70e2611fb7105bbc711/synapse/storage/databases/main/stream.py#L633-L639).

As you say, it's fine if we use an old value here. The existing code does so already (making use of `room_max_stream_ordering` after taking potentially multiple seconds to process state deltas). I can however document this in a comment to ensure the order doesn't get flipped around.

MadLittleMods commented on changelog.d/18926.bugfix:
```diff
@@ -0,0 +1 @@
+Fix performance regression related to the delayed events feature.
```
We should mention the MSC. I assume it's this one? https://github.com/matrix-org/matrix-spec-proposals/pull/4140

"experimental"

MadLittleMods reviewed (commented):
I haven't look at whether this is a good solution holistically with the delayed events feature. But the change itself looks like an improvement to what was there before.

anoadragon453 committed:
Note 100 row limit in docstrings

anoadragon453 committed:
Note that order of fetching stream orderings is important

anoadragon453 committed:
link MSC4140 in newsfile

anoadragon453 committed:
More error handling

anoadragon453 committed:
Revert "Only schedule next delayed event once"

This reverts commit 9f1c7e32bbfe49f5ffd1407e7814268757417ffc.

This change caused delayed events to not be scheduled, thus in turn
causing the `TestDelayedEvents/delayed_.*_events_are_sent_on_timeout`
tests to fail.

I don't fully understand why my change didn't work, but have spent
enough time looking at the code. The optimisation was very slight
anyhow, so I don't think it's worth persuing for longer.

MadLittleMods commented on synapse/handlers/delayed_events.py:
```diff
@@ -251,7 +251,23 @@ async def _handle_state_deltas(self, deltas: List[StateDelta]) -> None:
             )
 
-            sender = UserID.from_string(event_id_and_sender_dict[delta.event_id])
+            sender_str = event_id_and_sender_dict.get(delta.event_id, None)
+            if sender_str is None:
+                logger.debug(
+                    "Not handling delta for event where sender is unknown: %s",
+                    delta.event_id,
+                )
+                continue
```
For my own reference, when does this kind of thing happen?

Because we're working with `current_state_delta_stream`, this will be the case in state resets where the `delta.event_id` is already `None`. There are a few things here:

1. We should probably handle `delta.event_id` is `None` separately from a failed lookup which would be totally unexpected.
1. It seems like we should also cancel delayed state events in state reset cases.

anoadragon453 replied:
> Perhaps more what I'm saying is that we need to handle the case where `delta.event_id` will be `None`.

It seems what you're saying is that in the case of a state reset, `delta.event_id` is expected to be `None`. We currently handle this by continuing and stating that it's unexpected -- but it *is* expected in the case of state resets.

And it's possible to verify whether this is a state reset in a manner similar to how Sliding Sync does it. Is that correct?

MadLittleMods replied:
`delta.event_id` is expected to be `None` in a few scenarios (see the [linked comment](https://github.com/element-hq/synapse/blob/4367fb2d078c52959aeca0fe6874539c53e8360d/synapse/storage/databases/main/stream.py#L977-L981) above).

It seems like we should be handling these scenarios rather than skipping.

anoadragon453 replied:
Right, I think I understand you now. Does https://github.com/element-hq/synapse/pull/18926/commits/af7962dc0d12b9607dcc82a3054812d972a52250 better address this point?

MadLittleMods replied:
Perhaps it's trying to convey that since everyone on the local server has left, there is nothing more for us to do so skip is appropriate.

And we should differentiate the state reset scenario here to handle correctly.

I think the wording just needs to be adjusted.

MadLittleMods replied:
Updated in https://github.com/element-hq/synapse/pull/18926/commits/8c427f295f9a55e866cdac14abda9b354d8ee18c :white_check_mark:

MadLittleMods commented on synapse/handlers/delayed_events.py:
```diff
@@ -251,7 +251,23 @@ async def _handle_state_deltas(self, deltas: List[StateDelta]) -> None:
+                    sender_str,
+                    e,
+                )
+                continue
```
Feels like something that would totally break our assumptions in the codebase and we should blow up instead of continuing.

anoadragon453 replied:
I agree, though things such as [internal Matrix room link](https://matrix.to/#/!SGNQGPGUwtcPBUotTL:matrix.org/$ZHohYrFQWM9b3rFeWO7qWN5RSks4Osb-mgfoTBBXm40?via=jki.re&via=matrix.org&via=element.io) make it clear that we can't trust this data even today.

MadLittleMods commented on synapse/handlers/delayed_events.py:
```diff
@@ -146,10 +146,37 @@ async def process() -> None:
+
         # If self._event_pos is None then means we haven't fetched it from the DB yet
         if self._event_pos is None:
             self._event_pos = await self._store.get_delayed_events_stream_pos()
```
The `self._event_pos == room_max_stream_ordering` check below is ripe for getting us stuck in an infinite loop.

I think it should probably be `>=` 

We already mention that we can skip over some numbers in another spot:

https://github.com/element-hq/synapse/blob/4367fb2d078c52959aeca0fe6874539c53e8360d/synapse/handlers/delayed_events.py#L153-L155

anoadragon453 committed:
Prevent potential infinite loop if we bypass the max

Apply the same fix to a similar loop in the presence handling code.

anoadragon453 committed:
logger.error for missing sender

anoadragon453 committed:
state reset TODO

anoadragon453 committed:
Add a test for `get_senders_for_event_ids`

anoadragon453 committed:
Handle the case where a room has been purged

If a room has been purged, then the associated event would be gone as
well. Thus we wouldn't be able to pull out the sender.

This is different from the `sender` field being null on an event.

anoadragon453 committed:
Update logging and TODO to discuss state resets

anoadragon453 committed:
Move `Sentinel` to util module

anoadragon453 committed:
Replace ad-hoc Sentinel usage with utility module

anoadragon453 committed:
Update logs and comments to better cover cases
